cpp/lib/src/LocalDataDB.cc

```diff
@@ -0,0 +1,219 @@
+/** \file   LocalDataDB.cc
+ *  \brief  Implementation of the LocalDataDB class.
+ *  \author Dr. Johannes Ruscheinski (johannes.ruscheinski@uni-tuebingen.de)
+ *
+ *  \copyright 2020 Universitätsbibliothek Tübingen.  All rights reserved.
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Affero General Public License as
+ *  published by the Free Software Foundation, either version 3 of the
+ *  License, or (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "LocalDataDB.h"
+#include "Compiler.h"
+#include "DbConnection.h"
+#include "StringUtil.h"
+#include "UBTools.h"
+
+
+LocalDataDB::LocalDataDB(const OpenMode open_mode) {
+    db_connection_ = new DbConnection(UBTools::GetTuelibPath() + "local_data.sq3" /* must be the same path as in fetch_marc_updates.py */,
+                                      (open_mode == READ_WRITE) ? DbConnection::CREATE : DbConnection::READONLY);
+    if (open_mode == READ_ONLY)
+        return;
+
+    db_connection_->queryOrDie("CREATE TABLE IF NOT EXISTS local_data ("
+                               "    title_ppn TEXT PRIMARY KEY,"
+                               "    local_fields BLOB NOT NULL"
+                               ") WITHOUT ROWID");
+    db_connection_->queryOrDie("CREATE UNIQUE INDEX IF NOT EXISTS local_data_index ON local_data (title_ppn)");
+
+    db_connection_->queryOrDie("CREATE TABLE IF NOT EXISTS local_ppns_to_title_ppns_map ("
+                               "    local_ppn TEXT PRIMARY KEY,"
+                               "    title_ppn TEXT NOT NULL,"
+                               "    FOREIGN KEY(title_ppn) REFERENCES local_data(title_ppn)"
+                               ") WITHOUT ROWID");
+    db_connection_->queryOrDie("CREATE UNIQUE INDEX IF NOT EXISTS local_ppns_to_title_ppns_mapindex "
+                               "ON local_ppns_to_title_ppns_map (local_ppn)");
+}
+
+
+LocalDataDB::~LocalDataDB() {
+    delete db_connection_;
+}
+
+
+void LocalDataDB::clear() {
+    db_connection_->queryOrDie("DELETE * FROM local_data");
```

Comment:
* is not allowed in DELETE statements ("DELETE FROM local_data") should be correct for SQLite